File: algorithm/654.maximum-binary-tree.js
/*
 * @lc app=leetcode id=654 lang=javascript
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var constructMaximumBinaryTree = function (nums) {
  return createMaximumBinaryTree(nums, 0, nums.length - 1)
};

function createMaximumBinaryTree(nums, left, right) {
  if (left > right) {
    return null
  }

  const maxIndex = findMaxIndex(nums, left, right)
  const root = new TreeNode(nums[maxIndex],
    createMaximumBinaryTree(nums, left, maxIndex - 1),
    createMaximumBinaryTree(nums, maxIndex + 1, right))

  return root
}

function findMaxIndex(list, left, right) {
  let max = left

  for (let i = left + 1; i <= right; i++) {
    if (list[i] > list[max]) {
      max = i
    }
  }

  return max
}
// @lc code=end

File: algorithm/654.maximum-binary-tree.kt
/*
 * @lc app=leetcode id=654 lang=kotlin
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
  fun constructMaximumBinaryTree(nums: IntArray): TreeNode? {
    return createBinaryTree(nums, 0, nums.size - 1)
  }
  
  fun createBinaryTree(nums: IntArray, left: Int, right: Int): TreeNode? {
    if (left > right) {
    return null
    }
  
    val maxIndex = findMaxIndex(nums, left, right)
    val root = TreeNode(nums[maxIndex])
    root.left = createBinaryTree(nums, left, maxIndex - 1)
    root.right = createBinaryTree(nums, maxIndex + 1, right)
  
    return root
  }

  fun findMaxIndex(nums: IntArray, left: Int, right: Int): Int {
  var max = left
  
  for (i in left+1..right) {
    if (nums[i] > nums[max]) {
      max = i
    }
  }
    return max
  }
}
// @lc code=end

